### ISSUE TITLE
Restructure and extend the default member access inspections

### ISSUE THREAD

**Issue by MDoerner** [rubberduck-vba/Rubberduck#5155]
Currently, there are five inspections dealing with default member accesses: `ValueRequiredInspection`, `ProcedureRequiredInspection`, `DefaultMemberRequiredInspection` and `ObjectVariableRequiresSetInspection` dealing with failed default member resolutions and `ImplicitDefaultMemberAssignment` dealing with assignments to non-indexed default members.

I would like to propose to extend the coverage regarding successful default member accesses, but to give up the distinction between accesses and assignments in favor of another split. 

More precisely, I would like to propose to have different inspections for `UseOfBangNotation`, `ImplicitDefaultMemberAccess`, `IndexedDefaultMemberAccess` and `ProcedureDefaultMemberAccess` (This really deserves a warning!). The first three would further be split along usual accesses, recursive accesses and unbound accesses. This split will allow to assign separate severities: unbound accesses will be warnings, other implicit accesses suggestions and indexed accesses hints.

Is this sensible or does that produce too many separate inspections?

**Comment by bclothier** [comment]
Just to see if this would work. I did not follow completely on the part about splitting by usual, recursive and unbound....

| Inspection | Default Severity | Notes |
| --- | --- | --- |
| `UseOfBangNotation` | Warning |
| `ImplicitDefaultMemberAccess` | Warning |
| `IndexedDefaultMemberAccess` | Hint |
| `ProcedureDefaultMemberAccess` | Warning |
| `RecursiveDefaultMemberAccess` | Warning | May be shown with other inspection(s) |
| `UnboundDefaultMemberAccess` | Warning | May be shown with other inspection(s) | 
?

Also, I'm not sure how `ProcedureDefaultMemberAccess` is supposed to be different from `ImplicitDefaultMemberAccess`? Will the `ProcedureDefaultMemberAccess` only apply to the user's code where they've created a procedure as a default for their classes?

**Comment by retailcoder** [comment]
I think the distinction is useful and warranted, and the more inspections we have, and the more granular they are, the better IMO.

I'm not sure I understand though - does that mean we would be removing an inspection that explicitly calls out a missing `Set` keyword?

**Comment by MDoerner** [comment]
First, the second split would yield `ImplicitDefaultMemberAccess`, `ImplicitRecursiveDefaultMemberAccess` and `ImplicitUnboundDefaultMemberAccess`, etc.

Second, the `ProcedureDefaultMemberAccess` is for procedure coercions, i.e. if the default member is used because you used an object variable in a call statement instead of a procedure, function of property.

**Comment by bclothier** [comment]
I think in this particular case getting 2 inspections make more sense than getting 1 very specific inspection, IMO.

I also think `ProcedureDefaultMemberAccess` may need some clarification because that's not what I think of when I see code like `foo(bar)` (where `bar` is an object variable with a default member, and `foo` takes a value type for the parameter). `ImplicitDefaultMemberAccessInProcedureCall`, maybe? Too wordy, though.

**Comment by MDoerner** [comment]
I think you gat me wrong regarding procedure coersion. What I mean is what happens to `bar` in the following example.
```vba
Public Sub Foo()
    Dim bar As Class1
    Set bar = New Class1
    bar
End Sub
```
The last reference to `bar` is a call to the default member of `Class1`.

**Comment by MDoerner** [comment]
Closed via PR #5166
